In [1]:
import pandas as pd

In [2]:
from google.colab import drive
drive.mount('/gdrive')
%cd /gdrive/My Drive/Test

Mounted at /gdrive
/gdrive/My Drive/Test


In [3]:
test = pd.read_csv('us events ad_revenue filtered 03.02-07.02.csv')

In [4]:
test.head()

profile_id os_version device_manufacturer   device_model device_type  \
0  JLjpGzDHBQm         11             Samsung  Galaxy Tab S6      tablet   
1  JLjpGzDHBQm         11             Samsung  Galaxy Tab S6      tablet   
2  JLkVacrzHY6         11              Google        Pixel 2       phone   
3  JMC6wG3cX0f         10             Samsung      Galaxy S9       phone   
4  JLzWWTJfi1b         10             Samsung      Galaxy S9       phone   

  device_locale    event_name  \
0         en_US  c_ad_revenue   
1         en_US  c_ad_revenue   
2         en_US  c_ad_revenue   
3         en_US  c_ad_revenue   
4         en_US  c_ad_revenue   

                                          event_json       event_datetime  \
0  {"ad_type":"banner","placement":"","view_index...  2022-02-04 20:02:28   
1  {"ad_type":"banner","placement":"","view_index...  2022-02-04 20:02:45   
2  {"ad_type":"banner","placement":"","view_index...  2022-02-05 17:53:05   
3  {"ad_type":"int","placement":"next_level","vie...  2022-02-06 17:09:01   
4  {"ad_type":"banner","placement":"","view_index...  2022-02-06 17:15:38   

   event_timestamp event_receive_datetime  event_receive_timestamp  \
0       1643994148    2022-02-06 17:58:19               1644159499   
1       1643994165    2022-02-06 17:58:19               1644159499   
2       1644072785    2022-02-06 16:59:29               1644155969   
3       1644156541    2022-02-06 17:09:24               1644156564   
4       1644156938    2022-02-06 17:16:16               1644156976   

  connection_type operator_name country_iso_code       city  
0            wifi           NaN               US    Norwalk  
1            wifi           NaN               US    Norwalk  
2            wifi       Verizon               US  Charlotte  
3            wifi      T-Mobile               US      Cocoa  
4            cell      Verizon                US        NaN

In [5]:
df=test['event_json'].str.split(',')[0]
df

['{"ad_type":"banner"',
 '"placement":""',
 '"view_index":1',
 '"session_view_index":0',
 '"cur_ticks":637795729492245130',
 '"session_length":705.8807',
 '"revenue":0.000510313',
 '"ads_ltv":0.487876676518837',
 '"ecpm_sma":0.511508107185364',
 '"normalized_ecpm_sma":0.511508107185364',
 '"value":0.000510313',
 '"currency":"USD"',
 '"ad_country_code":"US"',
 '"network_name":"Google AdMob"',
 '"ad_unit_id":"64c6b1114ea8c1cc"',
 '"ad_network_placement":"ca-app-pub-7511632782617496/8825263691"',
 '"creativeId":"CJ3XirfD5vUCFYPnKAUdG3MOYQ"}']

In [6]:
import json

In [7]:
#распарсим столбец event_json
test_df = pd.io.json.json_normalize(test.event_json.apply(json.loads))

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  


In [8]:
#соединим два df
final_df = pd.concat([test, test_df], axis=1)

In [9]:
final_df.head()

profile_id os_version device_manufacturer   device_model device_type  \
0  JLjpGzDHBQm         11             Samsung  Galaxy Tab S6      tablet   
1  JLjpGzDHBQm         11             Samsung  Galaxy Tab S6      tablet   
2  JLkVacrzHY6         11              Google        Pixel 2       phone   
3  JMC6wG3cX0f         10             Samsung      Galaxy S9       phone   
4  JLzWWTJfi1b         10             Samsung      Galaxy S9       phone   

  device_locale    event_name  \
0         en_US  c_ad_revenue   
1         en_US  c_ad_revenue   
2         en_US  c_ad_revenue   
3         en_US  c_ad_revenue   
4         en_US  c_ad_revenue   

                                          event_json       event_datetime  \
0  {"ad_type":"banner","placement":"","view_index...  2022-02-04 20:02:28   
1  {"ad_type":"banner","placement":"","view_index...  2022-02-04 20:02:45   
2  {"ad_type":"banner","placement":"","view_index...  2022-02-05 17:53:05   
3  {"ad_type":"int","placement":"next_level","vie...  2022-02-06 17:09:01   
4  {"ad_type":"banner","placement":"","view_index...  2022-02-06 17:15:38   

   event_timestamp  ...   ads_ltv   ecpm_sma normalized_ecpm_sma     value  \
0       1643994148  ...  0.487877   0.511508            0.511508  0.000510   
1       1643994165  ...  0.488387   0.511389            0.511389  0.000510   
2       1644072785  ...  0.678479   0.928597            0.928597  0.002386   
3       1644156541  ...  0.013894  13.782633           13.893784  0.013894   
4       1644156938  ...  0.060571   1.660282            1.660282  0.006828   

  currency ad_country_code       network_name        ad_unit_id  \
0      USD              US       Google AdMob  64c6b1114ea8c1cc   
1      USD              US       Google AdMob  64c6b1114ea8c1cc   
2      USD              US  APPLOVIN_EXCHANGE  64c6b1114ea8c1cc   
3      USD              US           AppLovin  3ab7705b07ebb75f   
4      USD              US       Google AdMob  64c6b1114ea8c1cc   

                     ad_network_placement                  creativeId  
0  ca-app-pub-7511632782617496/8825263691  CJ3XirfD5vUCFYPnKAUdG3MOYQ  
1  ca-app-pub-7511632782617496/8825263691  CNa-lL_D5vUCFZrpKAUdU2QBDQ  
2                          banner_regular                  pandora-58  
3                           inter_regular                    13045364  
4  ca-app-pub-7511632782617496/6197324251  CMTi5e-h6_UCFUdVwQodt9EIMw  

[5 rows x 33 columns]

In [10]:
final_df.drop(columns=['event_json'], axis=1)

profile_id os_version device_manufacturer         device_model  \
0       JLjpGzDHBQm         11             Samsung        Galaxy Tab S6   
1       JLjpGzDHBQm         11             Samsung        Galaxy Tab S6   
2       JLkVacrzHY6         11              Google              Pixel 2   
3       JMC6wG3cX0f         10             Samsung            Galaxy S9   
4       JLzWWTJfi1b         10             Samsung            Galaxy S9   
...             ...        ...                 ...                  ...   
860074  JLSQY2nx1NQ         12             Samsung  Galaxy S21 Ultra 5G   
860075  JLKWwju79CA         12             Samsung             SM-G991U   
860076  JLKWwju79CA         12             Samsung             SM-G991U   
860077  JLKWwju79CA         12             Samsung             SM-G991U   
860078  JLKWwju79CA         12             Samsung             SM-G991U   

       device_type device_locale    event_name       event_datetime  \
0           tablet         en_US  c_ad_revenue  2022-02-04 20:02:28   
1           tablet         en_US  c_ad_revenue  2022-02-04 20:02:45   
2            phone         en_US  c_ad_revenue  2022-02-05 17:53:05   
3            phone         en_US  c_ad_revenue  2022-02-06 17:09:01   
4            phone         en_US  c_ad_revenue  2022-02-06 17:15:38   
...            ...           ...           ...                  ...   
860074       phone         en_US  c_ad_revenue  2022-02-03 08:29:27   
860075       phone         en_US  c_ad_revenue  2022-02-03 02:12:51   
860076       phone         en_US  c_ad_revenue  2022-02-03 02:13:06   
860077       phone         en_US  c_ad_revenue  2022-02-03 02:13:17   
860078       phone         en_US  c_ad_revenue  2022-02-03 03:11:44   

        event_timestamp event_receive_datetime  ...   ads_ltv   ecpm_sma  \
0            1643994148    2022-02-06 17:58:19  ...  0.487877   0.511508   
1            1643994165    2022-02-06 17:58:19  ...  0.488387   0.511389   
2            1644072785    2022-02-06 16:59:29  ...  0.678479   0.928597   
3            1644156541    2022-02-06 17:09:24  ...  0.013894  13.782633   
4            1644156938    2022-02-06 17:16:16  ...  0.060571   1.660282   
...                 ...                    ...  ...       ...        ...   
860074       1643866167    2022-02-03 08:29:37  ...  1.007506   0.080914   
860075       1643843571    2022-02-03 02:14:48  ...  0.277541   3.379858   
860076       1643843586    2022-02-03 02:14:48  ...  0.280321   3.319878   
860077       1643843597    2022-02-03 02:14:48  ...  0.283101   3.265896   
860078       1643847104    2022-02-03 21:12:34  ...  0.285881   3.217312   

       normalized_ecpm_sma     value currency ad_country_code  \
0                 0.511508  0.000510      USD              US   
1                 0.511389  0.000510      USD              US   
2                 0.928597  0.002386      USD              US   
3                13.893784  0.013894      USD              US   
4                 1.660282  0.006828      USD              US   
...                    ...       ...      ...             ...   
860074            0.080914  0.000020      USD              US   
860075            3.379858  0.006740      USD              US   
860076            3.319878  0.002780      USD              US   
860077            3.265896  0.002780      USD              US   
860078            3.217312  0.002780      USD              US   

             network_name        ad_unit_id  \
0            Google AdMob  64c6b1114ea8c1cc   
1            Google AdMob  64c6b1114ea8c1cc   
2       APPLOVIN_EXCHANGE  64c6b1114ea8c1cc   
3                AppLovin  3ab7705b07ebb75f   
4            Google AdMob  64c6b1114ea8c1cc   
...                   ...               ...   
860074  APPLOVIN_EXCHANGE  64c6b1114ea8c1cc   
860075       Google AdMob  64c6b1114ea8c1cc   
860076       Google AdMob  64c6b1114ea8c1cc   
860077       Google AdMob  64c6b1114ea8c1cc   
860078       Google AdMob  64c6b1114ea8c1cc 

In [11]:
final_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 860079 entries, 0 to 860078
Data columns (total 33 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   profile_id               860065 non-null  object 
 1   os_version               860079 non-null  object 
 2   device_manufacturer      860079 non-null  object 
 3   device_model             860079 non-null  object 
 4   device_type              860079 non-null  object 
 5   device_locale            860079 non-null  object 
 6   event_name               860079 non-null  object 
 7   event_json               860079 non-null  object 
 8   event_datetime           860079 non-null  object 
 9   event_timestamp          860079 non-null  int64  
 10  event_receive_datetime   860079 non-null  object 
 11  event_receive_timestamp  860079 non-null  int64  
 12  connection_type          860079 non-null  object 
 13  operator_name            841200 non-null  object 
 14  coun

In [12]:
final_df.describe()

event_timestamp  event_receive_timestamp     view_index  \
count     8.600790e+05             8.600790e+05  860079.000000   
mean      1.644088e+09             1.644088e+09       2.049220   
std       1.058604e+05             1.055928e+05       9.682474   
min       1.643467e+09             1.643836e+09       1.000000   
25%       1.644017e+09             1.644018e+09       1.000000   
50%       1.644102e+09             1.644103e+09       1.000000   
75%       1.644177e+09             1.644177e+09       1.000000   
max       1.644268e+09             1.644268e+09     417.000000   

       session_view_index     cur_ticks  session_length        revenue  \
count       860079.000000  8.600790e+05   860079.000000  860079.000000   
mean             0.241736  6.377966e+17     5730.505424       0.002780   
std              1.743580  1.060356e+12    15342.189669       0.015129   
min              0.000000  6.377900e+17        2.208125       0.000000   
25%              0.000000  6.377959e+17      512.645300       0.000046   
50%              0.000000  6.377968e+17     1398.315000       0.000494   
75%              0.000000  6.377975e+17     4095.932000       0.000533   
max             74.000000  6.378005e+17   394117.600000       0.937196   

             ads_ltv       ecpm_sma  normalized_ecpm_sma          value  
count  860079.000000  860079.000000        860079.000000  860079.000000  
mean        1.417849       3.147110             3.227661       0.002780  
std         2.119035      15.973978            16.433888       0.015129  
min         0.000026       0.001014             0.001014       0.000000  
25%         0.268843       0.139942             0.139942       0.000046  
50%         0.682342       0.480991             0.480991       0.000494  
75%         1.599739       0.772006             0.772006       0.000533  
max        26.699762     908.876160           917.695618       0.937196

In [14]:
agg_func_describe = {'ecpm_sma': ['describe']}
final_df.groupby(['ad_type']).agg(agg_func_describe).round(2)

ecpm_sma                                                 
         describe                                                 
            count   mean    std   min    25%    50%    75%     max
ad_type                                                           
banner   815984.0   0.78   1.31  0.00   0.13   0.45   0.63   58.17
int       37613.0  42.64  47.58  0.34  15.36  24.56  50.75  714.30
rv         6482.0  71.90  77.62  0.55  27.16  37.59  80.06  908.88

In [15]:
pd.pivot_table(final_df, index=['ad_type'])

ads_ltv           cur_ticks   ecpm_sma  event_receive_timestamp  \
ad_type                                                                     
banner   1.426535  637796639933102336   0.780689             1.644089e+09   
int      1.178056  637796598920186368  42.635934             1.644086e+09   
rv       1.715841  637796623724818688  71.901899             1.644087e+09   

         event_timestamp  normalized_ecpm_sma   revenue  session_length  \
ad_type                                                                   
banner      1.644088e+09             0.780689  0.000744     5801.271824   
int         1.644084e+09            44.175671  0.036210     4401.937939   
rv          1.644086e+09            73.655323  0.065051     4531.362973   

         session_view_index     value  view_index  
ad_type                                            
banner             0.000000  0.000744    1.000000  
int                4.854465  0.036210   22.350810  
rv                 3.906356  0.065051   16.326134

**Наблюдения **

В представленных выше таблицах видно, что больше всего было показано баннерных объявлений (кол-во = 81598), вторая по популярности полноэкранная реклама (кол-во = 37613), на последнем месте - видео (кол-во = 6482). Если сравнивать eCPM, то показатель выше всего у видео-рекламы, что говорит о том именно с видео приходит наибольшая выручка. Этот показатель практически в 2 раза выше, чем у полноэкранной рекламы и практически в 10 раз выше, чем у баннерной. 

In [16]:
columns=final_df['os_version']
print(type(columns))

<class 'pandas.core.series.Series'>


In [17]:
columns1=columns.value_counts()
columns1

12       407587
11       209071
10       150871
9         66662
8.1.0      9249
8.0.0      8016
7.0        4057
7.1.2      1811
7.1.1      1342
5.0.2       709
6.0.1       704
Name: os_version, dtype: int64

In [18]:
columns2=final_df['city'].unique()
len(columns2)

921

In [19]:
columns3=final_df['device_manufacturer'].unique()
columns3

array(['Samsung', 'Google', 'TCL', 'LG', 'OnePlus', 'Motorola',
       'Teracube', 'Lenovo', 'T-Mobile', 'Verizon Wireless', 'Xiaomi',
       'Alcatel', 'Cricket', 'Blu', 'Kyocera', 'Maxwest', 'Nokia', 'FIH',
       'Dish', 'Schok', 'deejoy', 'AT&T', 'Zte', 'Redmi', 'Yulong',
       'Wiko', 'Huawei', 'Onn (Walmart)', 'AEEZO', 'Asus', 'JOYAR',
       'POCO', 'Ulefone', 'Razer'], dtype=object)

In [20]:
cell = final_df['city']
print(type(cell))

<class 'pandas.core.series.Series'>


In [21]:
cell1=cell.value_counts()[:10]
cell1

New York        17584
Minneapolis     12980
Houston         10137
Springfield     10135
Charlotte        9750
Chicago          8697
San Antonio      7912
Indianapolis     7126
Kansas City      7038
Gallatin         6573
Name: city, dtype: int64

In [22]:
import numpy as np
table = pd.pivot_table(final_df, 
               values='ecpm_sma', 
               index=['city', 'profile_id'],
               columns=['ad_type'],
               aggfunc=np.mean)
table

ad_type                   banner         int         rv
city       profile_id                                  
Aberdeen   JL0GsLmgQqj  0.337810   15.321452  18.886984
Abilene    JLpM9tznEPx       NaN  189.868790        NaN
           JLtCrVvg6nt       NaN   43.893208  45.250385
           JLzhypEFyJW  1.153814   32.689328  60.879490
Addison    JLxvUGWsVRL  0.489450         NaN        NaN
...                          ...         ...        ...
Yukon      JLkn9RCY77U  0.201054   17.767026  26.133963
Yuma       JLxW1T3bKjT  0.163977   13.283648        NaN
Zanesville JLk6DA4qnFe  0.203445   28.501600        NaN
           JLtKjn6xUJb  0.138852   24.621979  79.751428
           JMFzObeOOQf  0.375181    9.680847        NaN

[3477 rows x 3 columns]

In [23]:
table.query('city==["New York"]')[:10]

ad_type                 banner         int  rv
city     profile_id                           
New York JLwYuhy0zur  2.183456  203.665415 NaN
         JLxx58yvrVE  1.023592   16.853265 NaN
         JLvfQR21ryN  0.831582   54.392418 NaN
         JLv7r60Usv9  2.273083   20.359159 NaN
         JLyLvC6WQzx  0.012801         NaN NaN
         JLwRdFS5uiX  0.551425         NaN NaN
         JLmHUire7Jt  0.147796   15.278293 NaN
         JLpX1hhdpmD  0.697166         NaN NaN
         JMGCiSPFpiP  0.207170   14.872630 NaN
         JLtHaP6fBat  1.945072  133.000837 NaN

In [24]:
table2 = pd.pivot_table(final_df, 
               values='ecpm_sma', 
               index=['os_version', 'profile_id'],
               columns=['ad_type'],
               aggfunc=np.mean)
table2

ad_type                   banner         int          rv
os_version profile_id                                   
10         JL06OWhBYxn  0.332892         NaN         NaN
           JL0eRcJ7HUF  2.578720  155.197458  209.661621
           JL0u6mnAMTV  1.578300   37.016598         NaN
           JL2s010u5r5  0.299757   22.034080   28.164125
           JL3hghqjRD2  1.156718   90.632641  133.300400
...                          ...         ...         ...
9          JMHME556xNs  0.102531   15.260992         NaN
           JMHjstfO7QG  0.717077   37.690791         NaN
           JMJcF8eWLf5  0.413694   23.369940         NaN
           JMKCFREXspE  0.089956    6.126935    7.120764
           JMKnVH32dVb       NaN   11.243364         NaN

[4245 rows x 3 columns]

In [25]:
table2.query('os_version=="12"')[10:30]

ad_type                   banner         int          rv
os_version profile_id                                   
12         JL0vSVQq5fu  3.299551         NaN         NaN
           JL0vdGxNqiM  0.588559   81.937248         NaN
           JL13ZBmWsNa  0.748686   23.100569         NaN
           JL1B020ytz8  0.353587   40.517599         NaN
           JL1BpRUyfqY  0.704698   22.922560         NaN
           JL1FhbpFZ66  0.666778   12.967731         NaN
           JL1LJGbUKca  0.260160    9.457488         NaN
           JL1LfavPEXc  1.152465   57.618320         NaN
           JL2g2jUx5wC  0.603246   49.135179         NaN
           JL3f2gDriFA  3.458225   45.522209   38.698219
           JL3hAzGikWg  1.888316   42.307858   29.085815
           JL3tbiZQVp8  0.952382         NaN         NaN
           JL42Wn7Qo91  0.844349   36.527394         NaN
           JL49WTG4cqE  0.591561  192.695368         NaN
           JL4JWUUaEcg  2.138449   35.974646         NaN
           JL4KSxxKnjo  0.392208    7.156405  119.040001
           JL4Ofhf1x65  0.566171         NaN         NaN
           JL4OpMmc4we  3.482962   39.520836         NaN
           JL4Z15ZWKWR  0.737046    2.778115         NaN
           JL4ZRQr8LeF  0.753667  121.438725  210.006393

In [26]:
sheet = pd.pivot_table(final_df, 
               values='revenue', 
               index=['os_version', 'profile_id'],
               columns=['ad_type'],
               aggfunc=np.mean)
sheet

ad_type                   banner       int        rv
os_version profile_id                               
10         JL06OWhBYxn  0.000468       NaN       NaN
           JL0eRcJ7HUF  0.002511  0.167240  0.211700
           JL0u6mnAMTV  0.001669  0.031279       NaN
           JL2s010u5r5  0.000300  0.022007  0.029616
           JL3hghqjRD2  0.001160  0.155709  0.211700
...                          ...       ...       ...
9          JMHME556xNs  0.000090  0.015434       NaN
           JMHjstfO7QG  0.000668  0.038351       NaN
           JMJcF8eWLf5  0.000384  0.023706       NaN
           JMKCFREXspE  0.000078  0.005683  0.007178
           JMKnVH32dVb       NaN  0.011334       NaN

[4245 rows x 3 columns]

In [27]:
sheet.query('os_version=="12"')[:10]

ad_type                   banner       int  rv
os_version profile_id                         
12         JL015rw9vmN  0.001161  0.021983 NaN
           JL06b4k4ygp  0.000437  0.007327 NaN
           JL09tSq5vCU  0.000510  0.016824 NaN
           JL0CA7Zu7Tp  0.000098  0.009117 NaN
           JL0Fx0jWp1n  0.000807       NaN NaN
           JL0PttDYfzH  0.000803  0.016855 NaN
           JL0hPQsuMML  0.005315  0.066480 NaN
           JL0nxvHc8s8  0.000969       NaN NaN
           JL0v0qUnWaD  0.002963       NaN NaN
           JL0vB2rdkeV  0.000776  0.021018 NaN

*Выводы*

Из представленных выше таблиц можно сделать вывод, что полноэкранная реклама дает более высокие показатели доходов и eCPM по сравнению с баннерной рекламой. Данных по видео-рекламе очень мало, что говорит о том, что видео вероятно не досматривают до конца. 



Также больше всего пользователей из таких городов, как Нью-Йорк(17584), Миннеаполис(12980) и Хьютон(10137). Самые используемые версии ОС - 12, 11 и 10. 

In [30]:
!git remote add origin https://<Uliliana>:<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>@github.com/<Uliliana>/reponame.git

/bin/bash: Uliliana: No such file or directory
